ci/clang_check_attributes.py

```diff
@@ -0,0 +1,285 @@
+#!/usr/bin/env python3
+#
+# Copyright 2022 by Antmicro. Verilator is free software; you
+# can redistribute it and/or modify it under the terms of either the GNU Lesser
+# General Public License Version 3 or the Apache License 2.0.
+# SPDX-License-Identifier: LGPL-3.0-only OR Apache-2.0
+#
+# Based on the work of (Guillaume "Vermeille" Sanchez)[https://github.com/Vermeille/clang-callgraph], licensed under Apache 2.0
+
+from pprint import pprint
+from clang.cindex import CursorKind, Index
+import sys
+import os
+import glob
+from termcolor import colored
+
+INDEX = None
+CLANG_ARGS = []
+TRANSLATE_UNITS = {}
+PRINTED = []
+
+# by definition constructor is MT safe,
+# but functions that constructor call are not!
+MT_SAFE_KINDS = [CursorKind.CONSTRUCTOR]
+UNSAFE_CALLS = []
+
+
+def get_diag_info(diag):
+    return {
+        'severity': diag.severity,
+        'location': diag.location,
+        'spelling': diag.spelling,
+        'ranges': list(diag.ranges),
+        'fixits': list(diag.fixits)
+    }
+
+
+def print_node(node, level):
+    if fully_qualified_pretty(node) not in PRINTED:
+        annotations = get_annotations(node)
+        if "MT_SAFE" in annotations or node.kind in MT_SAFE_KINDS:
```

Comment:
Please also look for VL_PURE.  Pure means a function has no external dependencies and is thus also VL_MT_SAFE.

Please check a VL_PURE function can only call another VL_PURE function not one that is VL_MT_SAFE.